retry/retry.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2024 Google LLC
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     https://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package retry
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+// IsRetriable is method signature for implementing to override default logic of retrying each error.
+type IsRetriable func(error) bool
+
+// Policy represents the struct to configure the retry behavior.
+type Policy struct {
+	// MaxAttempts represents the maximum number of retry attempts.
+	MaxAttempts int
+	// BackoffFactor is the multiplier by which retry interval (Jitter) increases after each retry.
+	// For constant backoff set Backoff factor to 1.
+	BackoffFactor float64
+	// Jitter is the interval before the first retry.
+	Jitter time.Duration
+	// ShouldRetry is the method to override default retry logic of retry every error.
+	ShouldRetry IsRetriable
+}
+
+// backoff computes interval between retries. Interval is jitter*(backoffFactor^attempt).
+// For e.g. if jitter was set to 10 and factor was 3, backoff between attempts would be [10, 30, 90, 270...].
+func backoff(attempt int, policy Policy) time.Duration {
+	b := float64(policy.Jitter) * math.Pow(policy.BackoffFactor, float64(attempt))
+	return time.Duration(b)
+}
+
+// isRetriable checks if error is retriable. If ShouldRetry is unimplemented it always returns
+// true, otherwise overriden method's logic determines the retry behavior.
+func isRetriable(policy Policy, err error) bool {
+	if policy.ShouldRetry == nil {
+		return true
+	}
+	return policy.ShouldRetry(err)
+}
+
+// RunWithResponse executes and retries the function on failure based on policy defined and returns response on success.
+func RunWithResponse[T any](ctx context.Context, policy Policy, f func() (T, error)) (T, error) {
```

Comment:
I believe you should expand the behavior of `f`. The way I read is "if f returns an error" then we'll retry; finish otherwise" or something like that. Also if I'm understanding it correctly why not return either `(T, bool, error)` where T is response type, bool tells if it should retry and error an actual error - in error case we simply abort and return the error.

Reply:
ah! what if `f` is nil?

Reply:
I did not add that check earlier as its for our internal use, we don't expect any bad inputs here. Updated for precautions.

Reply:
I don't want `f` to encapsulate the logic of whether error should be retried or not, that's the helper functions responsibility. We also have a way for caller to specify `shouldRetry()` logic.